### PR TITLE
ptr::eq: clarify that comparing dyn Trait is fragile

### PR DESCRIPTION
**Pull request by RalfJung**
Also remove the dyn trait example from `ptr::eq` since those tests are not actually guaranteed to pass due to how unstable vtable comparison is.

Cc @rust-lang/libs-api 
Cc discussion following https://github.com/rust-lang/rust/pull/80505

### CODE REVIEW

**Review by dtolnay — APPROVED** [review]
Thanks!

**Comment by RalfJung** [comment]
Looking at https://github.com/rust-lang/rust/pull/59390 which introduced these examples, we should probably have a line explicitly stating the wide pointer behavior. So I added that. Sorry for the late addition.

@bors r-
r? @dtolnay

**Comment by rib** [comment]
I wonder if some of the documentation changes here for `Arc` could do with further clarification and it's possible even that these changes make the semantics for `Arc::ptr_eq` more ambiguous now.

This PR references https://github.com/rust-lang/rust/pull/80505 which notably [appeared to agree](https://github.com/rust-lang/rust/pull/80505#issuecomment-757141449) that `Arc::ptr_eq` should return true when two Arcs point to the same allocation that's owned by the same counters.

In other words `Arc::ptr_eq` shouldn't conceptually compare the pointers to the wrapped values and so the warning about how `dyn Trait` wide pointers are compared may be misplaced and instead throws into doubt what "allocation" refers to.

Since pull request #80505  was closed (I think just because the patch wasn't updated according to feedback) I recently opened #103763 to keep track of the original issue about the `Arc::ptr_eq` semantics.

If the same conclusion as from [#80505](https://github.com/rust-lang/rust/pull/80505#issuecomment-757141449) is confirmed for #103763 then it would probably make sense to further iterate the Arc and Rc documentation to somehow clarify that the `ptr::eq` semantics for `dyn Trait` comparisons isn't something that should be a concern for users of `Arc::ptr_eq` or `Rc::ptr_eq`.

On the other hand if it's decided that `Arc::ptr_eq` is supposed to compare the wrapped value pointers (including their meta data) then it could also make sense to explicitly state that it is equivalent to `Arc::as_ptr(&a) == Arc::as_ptr(&b)` to try remove the uncertainty about what "allocation" refers to and why the semantics of `dyn Trait` wide pointer comparisons is relevant.

File: library/core/src/ptr/mod.rs [hunk 1]
```diff
@@ -1733,6 +1733,12 @@ pub(crate) unsafe fn align_offset<T: Sized>(p: *const T, a: usize) -> usize {
 /// by their address rather than comparing the values they point to
 /// (which is what the `PartialEq for &T` implementation does).
 ///
+/// When comparing wide pointers, both the address and the metadata are tested for equality.
+/// However, note that comparing trait object pointers (`*const dyn Trait`) is unrealiable: pointers
```
Comment (mzji) [inline library/core/src/ptr/mod.rs:1737]: typo: unrealiable -> unreliable

Comment (RalfJung) [reply library/core/src/ptr/mod.rs:1737]: Good catch! https://github.com/rust-lang/rust/pull/105474

**Pull request merged**